### PR TITLE
Removal of superfluous block inputs and outputs.

### PR DESCRIPTION
**Pull request by lvella**
This PR is already complex enough, so I think we should wrap it up.

There are a couple of related TODOs that I think should be done in a follow-up PR, namely, carry over outputs to be removed from the previous iteration to the next. This will advance part of the job done by the next pass (dangling removal) to this pass, enabling further IO removal.

### CODE REVIEW

File: src/utils/mod.rs [hunk 1]
```diff
@@ -1,3 +1,107 @@
 pub mod range_consolidation;
 pub mod rev_vec_filler;
 pub mod tree;
+
+/// Removes the values at the given sorted indices from a Vec,
+/// preserving the order of the remaining elements.
+///
+/// It is a little better than doing with `Vec::retain()` because
```
[the remark this answers is not in the document]
Comment (lvella) [reply src/utils/mod.rs]: This function could be implemented using retain, but they are not equivalent, and the average case performance would be worse.

**Comment by leonardoalt** [comment]
> afaiu, looks good to me. Since this is quite involved, I think it might be worth to have some unit tests over "manually built" dags here (for future PRs) to test normal and corner cases being handled by the pass

Makes sense. Also fwiw I ran the womir-openvm e2e tests including Reth and Geth on this branch and they pass.
